### PR TITLE
Setup initial build system

### PR DESCRIPTION
**Pull request by levithomason**
Here is an initial build.  Includes build, lint, and watching.  The package is built to `lib` when user's install it.  I saved code refactors for another PR, so we're building ES5 to ES5 here, for now.

### `npm run dev`

This is the primary development script.  It starts watchers for the build and lint scripts.  Lint errors do not interrupt the build and vice versa.  They run in independent subshells.

### All Scripts

```
prebuild          # cleans build before each build
build             # run babel
build:watch       # run babel in watch mode
dev               # build and lint in watch mode
lint              # run eslint
lint:fix          # run eslint with --fix
lint:watch        # run eslint in watch mode
postinstall       # when user's install shx, build it for them
test              # run ava, still
```

### Demo

1. start dev workflow, `npm run dev`
1. change `/src` with a syntax error, see lint errors and confirm build still runs
1. fix error, see updated lint output and confirm build ran

npm run dev

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+lib/*
```
Comment (levithomason) [inline .eslintignore]: Don't lint built assets.

File: .gitignore [hunk 2]
```diff
@@ -2,6 +2,7 @@
 npm-debug.log*
 node_modules
 tmp
+lib
```
Comment (levithomason) [inline .gitignore]: Don't commit built assets.

File: package.json [hunk 3]
```diff
@@ -2,8 +2,18 @@
   "name": "shx",
   "version": "0.0.1",
   "description": "Portable Shell Commands for Node",
-  "bin": { "shx" : "cli.js"},
+  "bin": {
+    "shx": "lib/cli.js"
+  },
```
Comment (levithomason) [inline package.json]: Link to the built cli, not the (soon to be) ES6/7 version.

File: package.json [hunk 4]
```diff
@@ -2,8 +2,18 @@
...
   "scripts": {
+    "prebuild": "rimraf lib",
+    "build": "babel src -d lib",
+    "build:watch": "npm run build -- -w",
+    "dev": "npm run build:watch & npm run lint:watch",
+    "lint": "eslint .",
+    "lint:fix": "npm run lint --silent -- --fix",
+    "lint:watch": "watch 'npm run lint --silent' src",
+    "postinstall": "npm run build",
```
Comment (levithomason) [inline package.json]: forgive me if i'm preaching to the choir here, for clarification:

- `rimraf` - cross platform `rm -rf`
- `--` in npm scripts is passing arguments from one npm script to the next
- `--silent` used in lint scripts to squelch `npm ERR` blocks (keeps lint errors clean on terminal)

File: package.json [hunk 5]
```diff
@@ -31,9 +41,12 @@
   "homepage": "https://github.com/shelljs/shx#readme",
   "devDependencies": {
     "ava": "^0.11.0",
+    "babel-cli": "^6.6.5",
+    "babel-preset-es2015": "^6.6.0",
     "eslint": "^1.10.3",
     "eslint-config-airbnb": "^5.0.1",
-    "eslint-plugin-react": "^3.16.1"
+    "eslint-plugin-react": "^3.16.1",
+    "watch": "^0.17.1"
```
Comment (levithomason) [inline package.json]: `watch` is great, but is a little slow.  They are using the default `watchFile` interval of 5007ms right now :S.  There is a [PR](https://github.com/mikeal/watch/pull/84) and I've pointed it out for fixing.  If they don't fix it soon, we can fork.

**Comment by nfischer** [comment]
@levithomason I saw some unexpected output from `npm install`

> ```
> $ npm install
> npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
> npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.8
> npm ERR! Linux 4.2.0-30-generic
> npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
> npm ERR! node v5.3.0
> npm ERR! npm  v3.5.2
> npm ERR! path /home/nate/PR_4/node_modules/shx/lib/cli.js
> npm ERR! code ENOENT
> npm ERR! errno -2
> npm ERR! syscall chmod
> 
> npm ERR! enoent ENOENT: no such file or directory, chmod '/home/nate/PR_4/node_modules/shx/lib/cli.js'
> npm ERR! enoent ENOENT: no such file or directory, chmod '/home/nate/PR_4/node_modules/shx/lib/cli.js'
> npm ERR! enoent This is most likely not a problem with npm itself
> npm ERR! enoent and is related to npm not being able to find a file.
> npm ERR! enoent
> 
> npm ERR! Please include the following file with any support request:
> npm ERR!     /home/nate/PR_4/node_modules/shx/npm-debug.log
> ```

I can confirm that `lib/cli.js` does not exist, since the `lib/` folder doesn't exist.

**Comment by levithomason** [comment]
aye aye, this will happen if you try to install from GitHub.  However, if you install from `npm` (once published) this will not be the case.

`react-router` seems to have solved this with a custom [`postinstall`](https://github.com/reactjs/react-router/blob/master/npm-scripts/postinstall.js) script that checks for the `lib` folder and instructs you install from `npm` rather than github if it is missing.

I have a couple ideas aside from the custom postinstall script.  I'll do some testing and let you know.

**Comment by nfischer** [comment]
I think what cash does to solve this is they just commit the equivalent of the `lib/` folder

**Comment by levithomason** [comment]
That is definitely easy :)  I think it is nicer to have clean PRs and git history and let CI/tooling handle and build/deploy stuff if possible.  If you're OK committing build artifacts we can simply not ignore `lib`.  As this project grows, or considering ShellJS, the build changes will be pretty hefty compared to source code changes though.

Another approach I've seen is to have the build version on a separate branch, just for dist.  Then, only CI pushes to it on successful release builds.

LMK what you'd like to do here.

***

>This is a known [npm issue](https://github.com/npm/npm/issues/3055) and there are some PRs open to fix it, but who knows how long until they are merged.

**Comment by nfischer** [comment]
@levithomason I'm in favor of including build artifacts in this case. It's nice because it makes it easy to run `npm install` to install all dependencies (which is my first instinct when collaborating on a project).

**Comment by nfischer** [comment]
Other than that issue, I really like this PR. @ariporad what are your thoughts?

File: package.json [hunk 6]
```diff
@@ -2,8 +2,18 @@
...
   "description": "Portable Shell Commands for Node",
-  "bin": { "shx" : "cli.js"},
+  "bin": {
+    "shx": "lib/cli.js"
+  },
   "scripts": {
+    "prebuild": "rimraf lib",
+    "build": "babel src -d lib",
+    "build:watch": "npm run build -- -w",
+    "dev": "npm run build:watch & npm run lint:watch",
```
Comment (ariporad) [inline package.json]: Suggestion: use http://npm.im/concurrently, for cross platform compatability.

**Comment by levithomason** [comment]
To be clear, `npm install` should run just fine, even without the `lib`:

npm i

You should only see this issue when doing...
```
$ npm install levithomason/shx#feature/setup-build
```

...because this ^ command does not first run the postinstall hooks to create the `lib` build.  I also tested with `rm -rf node_modules` and it works when installing deps.

Are you getting some other kind of error?

**Comment by nfischer** [comment]
I'm using a shelljs script I wrote. I'm intending to actually release this as OSS, but here's a gist: https://gist.github.com/nfischer/4fe3b67a54a6197824eb

**Comment by ariporad** [comment]
@nfischer: I'd rather not commit build artifacts. As long as building works on travis, I think that's fine. If we're in a pinch, we could force commit them to a branch.

**Comment by ariporad** [comment]
Other than that, this looks fantastic, that's so much!

I have one small nit, otherwise, LGTM!

**Comment by levithomason** [comment]
Added concurrently, thanks for the tip!

**Comment by levithomason** [comment]
Good to merge then @nfischer ?

**Comment by ariporad** [comment]
@levithomason: LGTM! I'm going to merge it manually (and squash your commits)?

**Comment by levithomason** [comment]
Go for it, squashed.

**Comment by ariporad** [comment]
Ok, I merged this into master manually (and rebased it at the same time), so I'm going to close this PR.

**Comment by nfischer** [comment]
Not sure what happened. This worked well for me before, but now I can't run anything anymore. `lib/cli.js` requires `./parseArgs`, which doesn't exist inside the `lib/` folder.

**Pull request closed without merging**